Correct / recommended way of using `user_data`

After the merge of [this PR](https://github.com/apify/crawlee-python/pull/540) I receive type errors working with `user_data`. Consider this sample:

```python
import asyncio

from crawlee import Request
from crawlee._utils.urls import convert_to_absolute_url, is_url_absolute
from crawlee.configuration import Configuration
from crawlee.parsel_crawler import ParselCrawler, ParselCrawlingContext
from crawlee.router import Router

router = Router[ParselCrawlingContext]()


@router.default_handler
async def default_handler(context: ParselCrawlingContext) -> None:
    for category in context.selector.xpath(
        '//div[@class="side_categories"]//ul/li/ul/li/a'
    ):
        item = {"title": category.xpath("normalize-space()").get()}
        url = category.xpath("./@href").get()
        if url is not None:
            if not is_url_absolute(url):
                url = str(convert_to_absolute_url(context.request.url, url))
            request = Request.from_url(url, method="GET", label="detail")
            request.user_data["item"] = item  # <--- TYPE ERROR
            await context.add_requests([request])


@router.handler("detail")
async def detail_handler(context: ParselCrawlingContext) -> None:
    item = context.request.user_data["item"]
    item["results"] = context.selector.xpath("normalize-space(//form//strong[1])").get()  # <-- TYPE ERROR
    await context.push_data(item)


async def main() -> None:
    config = Configuration.get_global_configuration()
    config.persist_storage = False
    config.write_metadata = False
    crawler = ParselCrawler(request_handler=router)
    await crawler.run(["https://books.toscrape.com"])
    data = await crawler.get_data()
    print(data.items)


if __name__ == "__main__":
    asyncio.run(main())
```

Both in VS Code (with Pylance) and CLI (mypy) I get type errors on the highlighted spots. Mypy reports this:

```
./venv/bin/mypy main.py 
main.py:23: error: Incompatible types in assignment (expression has type "dict[str, str | None]", target has type "JsonValue")  [assignment]
main.py:30: error: Unsupported target for indexed assignment ("list[JsonValue] | dict[str, JsonValue] | str | bool | int | float | None")  [index]
main.py:30: error: No overload variant of "__setitem__" of "list" matches argument types "str", "str | None"  [call-overload]
main.py:30: note: Possible overload variants:
main.py:30: note:     def __setitem__(self, SupportsIndex, JsonValue, /) -> None
main.py:30: note:     def __setitem__(self, slice, Iterable[JsonValue], /) -> None
Found 3 errors in 1 file (checked 1 source file)
```

--- a/src/crawlee/_request.py
+++ b/src/crawlee/_request.py
@@ -13,14 +13,13 @@ from pydantic import (
     BeforeValidator,
     ConfigDict,
     Field,
-    JsonValue,
     PlainSerializer,
     PlainValidator,
     TypeAdapter,
 )
 from typing_extensions import Self
 
-from crawlee._types import EnqueueStrategy, HttpHeaders, HttpMethod, HttpPayload, HttpQueryParams
+from crawlee._types import EnqueueStrategy, HttpHeaders, HttpMethod, HttpPayload, HttpQueryParams, JsonSerializable
 from crawlee._utils.requests import compute_unique_key, unique_key_to_request_id
 from crawlee._utils.urls import extract_query_params, validate_http_url
 
@@ -59,7 +58,7 @@ class CrawleeRequestData(BaseModel):
     forefront: Annotated[bool, Field()] = False
 
 
-class UserData(BaseModel, MutableMapping[str, JsonValue]):
+class UserData(BaseModel, MutableMapping[str, JsonSerializable]):
     """Represents the `user_data` part of a Request.
 
     Apart from the well-known attributes (`label` and `__crawlee`), it can also contain arbitrary JSON-compatible
@@ -67,15 +66,15 @@ class UserData(BaseModel, MutableMapping[str, JsonValue]):
     """
 
     model_config = ConfigDict(extra='allow')
-    __pydantic_extra__: dict[str, JsonValue] = Field(init=False)  # pyright: ignore
+    __pydantic_extra__: dict[str, JsonSerializable] = Field(init=False)  # pyright: ignore
 
     crawlee_data: Annotated[CrawleeRequestData | None, Field(alias='__crawlee')] = None
     label: Annotated[str | None, Field()] = None
 
-    def __getitem__(self, key: str) -> JsonValue:
+    def __getitem__(self, key: str) -> JsonSerializable:
         return self.__pydantic_extra__[key]
 
-    def __setitem__(self, key: str, value: JsonValue) -> None:
+    def __setitem__(self, key: str, value: JsonSerializable) -> None:
         if key == 'label':
             if value is not None and not isinstance(value, str):
                 raise ValueError('`label` must be str or None')
@@ -139,7 +138,7 @@ class BaseRequestData(BaseModel):
     data: Annotated[dict[str, Any], Field(default_factory=dict)] = {}
 
     user_data: Annotated[
-        dict[str, JsonValue],  # Internally, the model contains `UserData`, this is just for convenience
+        dict[str, JsonSerializable],  # Internally, the model contains `UserData`, this is just for convenience
         Field(alias='userData', default_factory=lambda: UserData()),
         PlainValidator(user_data_adapter.validate_python),
         PlainSerializer(

--- a/src/crawlee/_types.py
+++ b/src/crawlee/_types.py
@@ -3,7 +3,7 @@ from __future__ import annotations
 from collections.abc import Iterator, Mapping
 from dataclasses import dataclass, field
 from enum import Enum
-from typing import TYPE_CHECKING, Annotated, Any, Literal, Protocol, Union
+from typing import TYPE_CHECKING, Annotated, Any, Literal, Protocol, TypeVar, Union
 
 from pydantic import ConfigDict, Field, PlainValidator, RootModel
 from typing_extensions import NotRequired, TypeAlias, TypedDict, Unpack
@@ -21,10 +21,20 @@ if TYPE_CHECKING:
     from crawlee.sessions._session import Session
     from crawlee.storages._dataset import ExportToKwargs, GetDataKwargs, PushDataKwargs
 
-# Type for representing json-serializable values. It's close enough to the real thing supported
-# by json.parse, and the best we can do until mypy supports recursive types. It was suggested
-# in a discussion with (and approved by) Guido van Rossum, so I'd consider it correct enough.
-JsonSerializable: TypeAlias = Union[str, int, float, bool, None, dict[str, Any], list[Any]]
+    # Workaround for https://github.com/pydantic/pydantic/issues/9445
+    J = TypeVar('J', bound='JsonSerializable')
+    JsonSerializable: TypeAlias = Union[
+        list[J],
+        dict[str, J],
+        str,
+        bool,
+        int,
+        float,
+        None,
+    ]
+else:
+    from pydantic import JsonValue as JsonSerializable
+
 
 HttpMethod: TypeAlias = Literal['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'TRACE', 'PATCH']
 

